profiling/src/profile/mod.rs

```diff
@@ -127,6 +127,127 @@ pub struct StackTraceId(usize);
+
+impl LabelSetId {
+    pub fn new<T>(v: T) -> Self
+    where
+        T: TryInto<u32>,
+        T::Error: Debug,
+    {
+        let index: u32 = v.try_into().expect("LabelSetId to fit into a u32");
+        Self(index)
+    }
+}
+
+impl From<LabelSetId> for u64 {
+    fn from(s: LabelSetId) -> Self {
+        Self::from(&s)
+    }
+}
+
+impl From<&LabelSetId> for u64 {
+    fn from(s: &LabelSetId) -> Self {
+        s.0.into()
+    }
+}
+
+impl From<LabelSetId> for usize {
+    fn from(s: LabelSetId) -> Self {
+        Self::from(&s)
+    }
+}
+
+impl From<&LabelSetId> for usize {
+    fn from(s: &LabelSetId) -> Self {
+        // Safety: we only run on platforms where usize >= u32
+        s.0 as usize
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash, PartialOrd, Ord)]
+#[repr(transparent)]
+/// NonZeroU32 allows niche optimization when this is stored in an Option
+/// Only access using the `new` and `get` functions.
+pub struct LabelId(NonZeroU32);
+
+impl LabelId {
+    pub fn new<T>(v: T) -> Self
+    where
+        T: TryInto<u32>,
+        T::Error: Debug,
+    {
+        let index: u32 = v.try_into().expect("LabelId to fit into a u32");
+
+        // PProf takes Labels by value
+        // But, keeping the type NonZeroU32 allows niche optimization when this is stored in an Option
+        let index = index.checked_add(1).expect("LabelId to fit into a u32");
+        // Safety: the `checked_add(1).expect(...)` guards this from ever being zero.
+        let index = unsafe { NonZeroU32::new_unchecked(index) };
+        Self(index)
+    }
+
+    pub fn get(&self) -> usize {
+        // -1 since we added 1 to make the index guaranteed nonzero in add_label
+        let i: usize = self.0.get().try_into().unwrap();
+        i - 1
+    }
+}
+
+impl From<LabelId> for u64 {
+    fn from(s: LabelId) -> Self {
+        Self::from(&s)
+    }
+}
+
+impl From<&LabelId> for u64 {
+    fn from(s: &LabelId) -> Self {
+        s.0.get().into()
+    }
+}
+
+/// A canonical representation for sets of labels.
+/// You should only use the impl functions to modify this.
+#[derive(Clone, Debug, Default, Eq, PartialEq, Hash)]
+pub struct LabelSet {
+    sorted_labels: Vec<LabelId>,
+}
+
+impl LabelSet {
+    pub fn new() -> Self {
+        Self {
+            sorted_labels: vec![],
+        }
+    }
+
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            sorted_labels: Vec::with_capacity(capacity),
+        }
+    }
+
+    pub fn from_vec(mut v: Vec<LabelId>) -> Self {
+        v.sort_unstable();
+        Self { sorted_labels: v }
+    }
+
+    pub fn iter(&self) -> core::slice::Iter<'_, LabelId> {
+        self.sorted_labels.iter()
+    }
+
+    pub fn add(&mut self, l: LabelId) {
+        self.sorted_labels.push(l);
+        self.sorted_labels.sort_unstable();
+    }
+
+    pub fn extend(&mut self, ls: &[LabelId]) {
+        self.sorted_labels.extend(ls);
+        self.sorted_labels.sort_unstable();
+    }
```

Comment:
I wonder if we could get by without allowing the `LabelSet` to grow. E.g, only allow it to be created with a fixed-size vec and that's it?